doc/sfp-cmis/CMIS-custom-SI-settings.md

```diff
@@ -0,0 +1,255 @@
+                          “lane5” : “9”,
+                          “lane6” : “9”,
+                          “lane7” : “9”,
+                      } 
+                  } 
+              } 
+          } 
+      }
+  }
+```
+
+# Proposed Work-Flow
+Please refer below points in line with flow diagram.
+
+1. When CMIS-supported module insertion happens in XCVRD, the module will progress to the AP_CONFIG state in the CMIS state machine. After which, when the module is in DataPathDeactivated or Disabled state, check if the OPTICS_SI_PARAM.json file is parsed successfully and if lane speed needs special Signal Integrity (SI) settings.
+
+2. If both of the above conditions are met, then proceed to generate the key (module key) and retrieve the SI attribute list.  
+  2.1. The JSON file has two directions (TX_SETTING and RX_SETTING) blocks. Each of these blocks contains sub-blocks of module vendors and other details that will help identify the best match to generate the SI attribute list, if applicable.  
+  2.2.  The EQ_FIXED flag is required to validate if the TX input EQ (equalization) setting is adaptive or fixed. TX EQ settings are adaptive by default and should be disabled to apply host-defined SI settings. Currently, fixed EQ settings are not recommended for TX input EQ settings. This kind of validation is not required to apply RX EQ settings. If EQ_FIXED is set to true only, then flow will move ahead with list generation, or else it will return an empty list for the direction (TX/RX) setting (2.10).   
+  2.3. Generate module key based on Data Path(DP) lane speed + module vendor name, + module vendor part number.    
+  2.4. Using this module key, the search begins for the detected port in the GLOBAL_MEDIA_SETTINGS section of the applicable direction setting block. If the module key matches entries or any search that matches for (2.5) to (2.8) sections, then SI attributes from this section are copied to the SI param attribute list (2.9).   
+  2.5. If no match happens in 2.4, reduce the search to module default + speed of the detected port in GLOBAL_MEDIA_SETTINGS  
+  2.6. If no match happens in 2.5, reduce the search to the speed of the detected port in GLOBAL_MEDIA_SETTINGS  
+  2.7. If no match happens in the GLOBAL_MEDIA_SETTINGS block, the search now begins in the PORT_MEDIA_SETTINGS block for the detected port. If no match happens in the PORT_MEDIA_SETTINGS block, the final search for the default block is done.  
+  2.8. If no match happens to the default block, then an empty attribute list (2.10) is returned.
+
+3. Get the attribute list and validate if the list is not empty, then proceed to process the SI setting param list. If the list is empty, continue with the DP_INIT state in the CMIS state machine.
+
+4. Apply the application code to the configuration with Explicit Control (EC) = 0, and commit to the activate state.  
+Reference Register: Upper Page 10h bytes 145 –152 (desired ApSel Code)
+
```

Comment:
The app code will be applied also from api.set_application(host_lanes_mask, appl) in CMIS_STATE_AP_CONF. Correct ?

Reply:
Yes. This is to read and cache module default SI setting.